src/executor/sort_merge_join.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+use std::vec::Vec;
+
+use futures::TryStreamExt;
+
+use super::*;
+use crate::array::{DataChunk, DataChunkBuilder, RowRef};
+use crate::types::DataType;
+
+/// The executor for sort merge join
+pub struct SortMergeJoinExecutor {
+    pub left_child: BoxedExecutor,
+    pub right_child: BoxedExecutor,
+    pub left_column_indexes: Vec<usize>,
+    pub right_column_indexes: Vec<usize>,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
+}
+#[allow(dead_code)]
+impl SortMergeJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    #[allow(dead_code)]
+    pub async fn execute(self) {
+        // collect all chunks from children
+        let (left_chunks, right_chunks) = async {
+            tokio::try_join!(
+                self.left_child.try_collect::<Vec<DataChunk>>(),
+                self.right_child.try_collect::<Vec<DataChunk>>(),
+            )
+        }
+        .await?;
+
+        // get rows iterator from chunks
+        let left_rows = || left_chunks.iter().flat_map(|chunk| chunk.rows()).peekable();
+        let right_rows = || {
+            right_chunks
+                .iter()
+                .flat_map(|chunk| chunk.rows())
+                .peekable()
+        };
+
+        // build
+        let data_types = self.left_types.iter().chain(self.right_types.iter());
+        let mut builder = DataChunkBuilder::new(data_types, PROCESSING_WINDOW_SIZE);
+        let mut left_row = left_rows().next().unwrap();
```

Comment:
I find 'left_rows().next().unwrap' didn't return the next row in rows. Is there a method other than for row in left_rows() can retun next row ? I found for method in [here](https://github.com/risinglightdb/risinglight/blob/a23b9b15337ff4aedcbbec9d638acb0e0562822e/src/executor/hash_join.rs#L56)

Reply:
left_rows() is an iterator. You may store it in advance.

```rust
let left_rows_iter = left_rows();
let first_row = left_rows_iter.next().unwrap();
```

By the way, unwrapping here is dangerous, as there might be no more rows.

Reply:
> By the way, unwrapping here is dangerous, as there might be no more rows.
Thanks for reminding me！I should be more careful